src/Vessel.cpp

```diff
@@ -0,0 +1,285 @@
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+BrewTroller is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with BrewTroller.  If not, see <http://www.gnu.org/licenses/>.
+
+
+BrewTroller - Open Source Brewing Computer
+
+Documentation, Forums and more information available at http://www.brewtroller.com
+*/
+
+
+/*
+The vessel class stores and manages data and actions pertaining to vessels, such as the HLT, MLT and kettle. It is a concrete class
+and should not be subclassed, for performance reasons and because it's important for other parts of the code to be able to interact
+with vessels without regard to the role they are playing (HLT vs. MLT vs. kettle) so that we can easily support different vessel configurations.
+*/
+
+#include <avr/eeprom.h>
+
+#include "Vessel.h"
+#include "EEPROM.h"	
+#include "Temp.h"
+#include "HardwareProfile.h"
+#include "Volume.h"
+#include "Outputs.h" //Needed only for minTriggerPin
+
+extern unsigned int cycleStart[4];
+extern bool estop;
+extern int temp[9];
+
+
+// set what the PID cycle time should be based on how fast the temp sensors will respond
+#if TS_ONEWIRE_RES == 12
+#define PID_CYCLE_TIME 750
+#elif TS_ONEWIRE_RES == 11
+#define PID_CYCLE_TIME 375
+#elif TS_ONEWIRE_RES == 10
+#define PID_CYCLE_TIME 188
+#elif TS_ONEWIRE_RES == 9
+#define PID_CYCLE_TIME 94
+#else
+// should not be this value, fail the compile
+#ERROR
+#endif
+//Note that the includeAux numbers should be addresses e.g. TS_AUX1, not just a number (that's why they're bytes instead of booleans).
+	Vessel::Vessel(byte initEepromIndex, byte initIncludeAux[], byte FFBias, float initMinVolume = 0, byte initMinTriggerPin = 0, byte initMaxPower = 100)
+	{
+
+		for (int i = 0; i < 10; i++)
+			volumeReadings[i] = 0;
+
+		eepromIndex = initEepromIndex;
+		minVolume = initMinVolume;
+		maxPower = initMaxPower;
+		minTriggerPin = initMinTriggerPin;
+		includeAux[0] = initIncludeAux[0];
+		includeAux[1] = initIncludeAux[1];
+		includeAux[2] = initIncludeAux[2];
+
+		usesAuxInputs = includeAux[0] + includeAux[1] + includeAux[2];
+
+		//Load settings from eeprom
+
+		//Load temperature sensor information - needs to be before PID setup so we have the temp probe addresses in place
+		//**********************************************************************************
+		//Temperature Sensors: HLT (0-7), MASH (8-15), KETTLE (16-23), H2OIN (24-31), H2OOUT (32-39),
+		//          BEEROUT (40-47), AUX1 (48-55), AUX2 (56-63), AUX3 (64-71)
+		//**********************************************************************************
+		EEPROMreadBytes(8 * eepromIndex, sensorAddress, 8);
```

Comment:
I agree. You mentioned that you had a good implementation of eeprom
encapsulation so I was planning to do all the eeprom interaction using
that, and figuring we would not make any eeprom schema changes until we had
better encapsulation. If you want to send me that implementation I'll
incorporate it into this project. Otherwise I can implement an
encapsulation class myself. Basically the vessel load code should be able
to call byte* EEPROM->fetchVesselTemperatureSensorID(VS_HLT), bool
fetchVesselUsesPID(VS_HLT), etc. Or we could do it as EEPROM->load(VS_HLT,
TEMP_SENSOR), that might be better.

I do think that the vessel class should be able to directly load its
initialization values itself, rather than some other function loading them
from eeprom and passing them to the class, which introduces opportunities
for errors and requires us to keep the eeprom load code and vessel code in
sync. Letting the vessel fetch values from eeprom lets us change the eeprom
schema and the vessel class independently, as long as we don't need to
add/cut values from eeprom.

Send over that encapsulation example and I'll incorporate it, or let me
know if you want me to write that myself.


On Thu, Dec 17, 2015 at 12:47 PM, Eric Yanush <notifications@github.com>
wrote:

> In src/Vessel.cpp
> <https://github.com/BrewTroller/BrewTroller-Official/pull/5#discussion_r47958786>
> :
>
> > +		maxPower = initMaxPower;
> > +		minTriggerPin = initMinTriggerPin;
> > +		includeAux[0] = initIncludeAux[0];
> > +		includeAux[1] = initIncludeAux[1];
> > +		includeAux[2] = initIncludeAux[2];
> > +
> > +		usesAuxInputs = includeAux[0] + includeAux[1] + includeAux[2];
> > +
> > +		//Load settings from eeprom
> > +
> > +		//Load temperature sensor information - needs to be before PID setup so we have the temp probe addresses in place
> > +		//**********************************************************************************
> > +		//Temperature Sensors: HLT (0-7), MASH (8-15), KETTLE (16-23), H2OIN (24-31), H2OOUT (32-39),
> > +		//          BEEROUT (40-47), AUX1 (48-55), AUX2 (56-63), AUX3 (64-71)
> > +		//**********************************************************************************
> > +		EEPROMreadBytes(8 * eepromIndex, sensorAddress, 8);
>
> By moving the eeprom reading and writing directly into the vessel class we
> limit the ability to maintain an obvious and cohesive data storage schema
> in the eeprom. Any changes to the layout of the storage schema in the
> eeprom.h/cpp files have the potential to break the reads and writes here.
>
> Although I am not a fan of the current eeprom storage implementation we
> use, I think that maintaining the storage schema in the eeprom
> implementation and have each client use its interface makes changes to the
> data schema or implementation transparent to the client, and is the right
> way to prevent fragility in persistent data storage.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BrewTroller/BrewTroller-Official/pull/5/files#r47958786>
> .
>